codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -383,59 +392,100 @@ private void writeGoAwayFrame(ChannelHandlerContext ctx, Http2GoAwayFrame frame,
-            // and so not correctly set the property for the buffered stream.
-            //
-            // See https://github.com/netty/netty/issues/8692
-            Object old = frameStreamToInitializeMap.put(streamId, stream);
-
-            // We should not re-use ids.
-            assert old == null;
+        }
+    }
 
-            encoder().writeHeaders(ctx, streamId, headersFrame.headers(), headersFrame.padding(),
-                    headersFrame.isEndStream(), promise);
+    private void writePushPromise(final ChannelHandlerContext ctx, Http2PushPromiseFrame pushPromiseFrame,
+                                  final ChannelPromise promise) {
+        if (isStreamIdValid(pushPromiseFrame.pushStream().id())) {
+            encoder().writePushPromise(ctx, pushPromiseFrame.stream().id(), pushPromiseFrame.pushStream().id(),
+                    pushPromiseFrame.http2Headers(), pushPromiseFrame.padding(), promise);
+        } else {
+            boolean init = initializeNewStream(ctx, (DefaultHttp2FrameStream) pushPromiseFrame.pushStream(), promise);
+            if (init) {
+                final int streamId = pushPromiseFrame.stream().id();
 
-            if (!promise.isDone()) {
-                numBufferedStreams++;
-                // Clean up the stream being initialized if writing the headers fails and also
-                // decrement the number of buffered streams.
-                promise.addListener(new ChannelFutureListener() {
-                    @Override
-                    public void operationComplete(ChannelFuture channelFuture) {
-                        numBufferedStreams--;
+                encoder().writePushPromise(ctx, streamId, pushPromiseFrame.pushStream().id(),
+                        pushPromiseFrame.http2Headers(), pushPromiseFrame.padding(), promise);
 
-                        handleHeaderFuture(channelFuture, streamId);
-                    }
-                });
-            } else {
-                handleHeaderFuture(promise, streamId);
+                if (promise.isDone()) {
+                    handleHeaderFuture(promise, streamId);
+                } else {
+                    numBufferedStreams++;
+                    // Clean up the stream being initialized if writing the headers fails and also
+                    // decrement the number of buffered streams.
+                    promise.addListener(new ChannelFutureListener() {
+                        @Override
+                        public void operationComplete(ChannelFuture channelFuture) {
+                            numBufferedStreams--;
+
+                            handleHeaderFuture(channelFuture, streamId);
+                        }
+                    });
+                }
             }
         }
     }
 
+    public boolean initializeNewStream(ChannelHandlerContext ctx, DefaultHttp2FrameStream http2FrameStream,
```

Comment:
So I strongly believe this should not be public... If the user needs this method we need to find a better way of exposing this

Reply:
@hyperxpro ^^